src/dune_engine/build_system.ml

```diff
@@ -2257,16 +2259,19 @@ let prefix_rules (prefix : unit Action_builder.t) ~f =
   in
   res
 
-let report_early_exn ~report_error exn =
-  let t = t () in
-  let error = { Error.exn; id = Error.Id.gen () } in
-  t.errors <- error :: t.errors;
+let report_early_exn ~report_error (exn : Exn_with_backtrace.t) =
   (match !Clflags.report_errors_config with
   | Early
   | Twice ->
     report_error exn
   | Deterministic -> ());
-  t.handler.error [ Add error ]
+  match exn.exn with
+  | Memo.Non_reproducible Scheduler.Run.Build_cancelled -> Fiber.return ()
```

Comment:
I don't think this should works because Memo drops the `Memo.Non_reproducible` wrapper when re-raising from what I remember.

Did you test this?